api/client/alpn_conn_upgrade.go

```diff
@@ -182,26 +185,34 @@ func (d alpnConnUpgradeDialer) DialContext(ctx context.Context, network, addr st
 
 	tlsConn := tls.Client(conn, cfg)
 
-	err = upgradeConnThroughWebAPI(tlsConn, url.URL{
+	upgradeURL := url.URL{
 		Host:   addr,
 		Scheme: "https",
 		Path:   constants.WebAPIConnUpgrade,
-	})
-	if err != nil {
-		defer tlsConn.Close()
-		return nil, trace.Wrap(err)
 	}
-	return tlsConn, nil
+
+	switch {
+	case d.withPing:
+		if err := upgradeConnThroughWebAPI(tlsConn, upgradeURL, constants.WebAPIConnUpgradeTypeALPNPing); err != nil {
+			return nil, trace.NewAggregate(err, tlsConn.Close())
+		}
+		return pingconn.New(tlsConn), nil
+
+	default:
+		if err := upgradeConnThroughWebAPI(tlsConn, upgradeURL, constants.WebAPIConnUpgradeTypeALPN); err != nil {
+			return nil, trace.NewAggregate(err, tlsConn.Close())
+		}
+		return tlsConn, nil
+	}
 }
 
-func upgradeConnThroughWebAPI(conn net.Conn, api url.URL) error {
+func upgradeConnThroughWebAPI(conn net.Conn, api url.URL, upgradeType string) error {
```

Comment:
can  we change the `upgradeConnThroughWebAPI` return param to  `(net.Conn, error`) and the return either conn or `pingconn.New(conn)`  depending on the type ?